File: VB/tagsList/Taglist/TagListButtonPainter.vb
' Developer Express Code Central Example:
' How to create a control that shows a list of tags
' 
' This example demonstrates how to create a TagList control editor that supports
' both standalone and in-place modes.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E4542


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.Utils.Drawing
Imports DevExpress.Skins
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.LookAndFeel
Imports DevExpress.Utils
Imports DevExpress.XtraEditors.ViewInfo
Imports System.Collections


Namespace tagsList

	Public Class TagListButtonPainter
		Inherits ButtonEditPainter
		Protected Overrides Sub DrawContent(ByVal info As ControlGraphicsInfoArgs)
			If (TryCast(info.ViewInfo, TagListViewInfo)).OwnerEdit Is Nothing Then
				MyBase.DrawButtons(info)
			Else
				TryCast((TryCast(info.ViewInfo, TagListViewInfo)).OwnerEdit, Taglist).Text = ""
				MyBase.DrawContent(info)
			End If
		End Sub
		Protected Overrides Overloads Sub DrawButtons(ByVal info As ControlGraphicsInfoArgs)
			MyBase.DrawButtons(info)
		End Sub

		Protected Overrides Sub DrawButton(ByVal viewInfo As ButtonEditViewInfo, ByVal info As EditorButtonObjectInfoArgs)
			MyBase.DrawButton(viewInfo, info)
			Dim vi As TagListViewInfo = TryCast(viewInfo, TagListViewInfo)
			Dim closeButtonRect As Rectangle = TagListViewInfo.GetCloseRectangle(info.Bounds)
			Dim currentButtonPainter As SkinEditorButtonPainter = CType(viewInfo.GetButtonPainter(info), SkinEditorButtonPainter)
			Dim currentSkin As Skin = EditorsSkins.GetSkin(currentButtonPainter.Provider)
			Dim closeButton As SkinElement = currentSkin(EditorsSkins.SkinCloseButton)
			Dim customSkinElementPainter As New CustomSkinElementPainter()
			If closeButton IsNot Nothing Then
				Dim skinInfo As New SkinElementInfo(EditorsSkins.GetSkin(currentButtonPainter.Provider)(EditorsSkins.SkinCloseButton), closeButtonRect)
				skinInfo.Cache = info.Cache
				skinInfo.State = info.State
				customSkinElementPainter.DrawObject(skinInfo)
			End If
		End Sub
	End Class
	Public Class CustomSkinElementPainter
		Inherits SkinElementPainter
		Public Overrides Overloads Sub DrawObject(ByVal e As ObjectInfoArgs)
			Dim ee As SkinElementInfo = TryCast(e, SkinElementInfo)
			If ee.Element Is Nothing Then
				Return
			End If
			DrawSkinForeground(ee)
		End Sub

		Protected Overrides Sub DrawSkinImage(ByVal elementInfo As SkinElementInfo, ByVal skinImage As SkinImage)
			MyBase.DrawSkinImage(elementInfo, skinImage)
		End Sub
		Public Sub PublicForeground(ByVal ee As SkinElementInfo)
			MyBase.DrawSkinForeground(ee)
		End Sub
	End Class
End Namespace

File: CS/tagsList/Taglist/TagListButtonPainter.cs
// Developer Express Code Central Example:
// How to create a control that shows a list of tags
// 
// This example demonstrates how to create a TagList control editor that supports
// both standalone and in-place modes.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E4542

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.Utils.Drawing;
using DevExpress.Skins;
using DevExpress.XtraEditors.Controls;
using DevExpress.LookAndFeel;
using DevExpress.Utils;
using DevExpress.XtraEditors.ViewInfo;
using System.Collections;


namespace tagsList {

    public class TagListButtonPainter : ButtonEditPainter {
        protected override void DrawContent(ControlGraphicsInfoArgs info) {
            if((info.ViewInfo as TagListViewInfo).OwnerEdit == null) {
                base.DrawButtons(info);
            }
            else {
                ((info.ViewInfo as TagListViewInfo).OwnerEdit as Taglist).Text = "";
                base.DrawContent(info);
            }
        }
        protected override void DrawButtons(ControlGraphicsInfoArgs info) {
            base.DrawButtons(info);
        }
        
        protected override void DrawButton(ButtonEditViewInfo viewInfo, EditorButtonObjectInfoArgs info) {
            base.DrawButton(viewInfo, info);
            TagListViewInfo vi = viewInfo as TagListViewInfo;
            Rectangle closeButtonRect = TagListViewInfo.GetCloseRectangle(info.Bounds);
            SkinEditorButtonPainter currentButtonPainter = (SkinEditorButtonPainter)viewInfo.GetButtonPainter(info);
            Skin currentSkin = EditorsSkins.GetSkin(currentButtonPainter.Provider);
            SkinElement closeButton = currentSkin[EditorsSkins.SkinCloseButton];
            CustomSkinElementPainter customSkinElementPainter = new CustomSkinElementPainter();
            if(closeButton != null) {
                SkinElementInfo skinInfo = new SkinElementInfo(EditorsSkins.GetSkin(currentButtonPainter.Provider)[EditorsSkins.SkinCloseButton], closeButtonRect);
                skinInfo.Cache = info.Cache;
                skinInfo.State = info.State;
                customSkinElementPainter.DrawObject(skinInfo);
            }
        }
    }
    public class CustomSkinElementPainter : SkinElementPainter {
        public override void DrawObject(ObjectInfoArgs e) {
            SkinElementInfo ee = e as SkinElementInfo;
            if(ee.Element == null) return;
            DrawSkinForeground(ee);
        }

        protected override void DrawSkinImage(SkinElementInfo elementInfo, SkinImage skinImage) {
            base.DrawSkinImage(elementInfo, skinImage);
        }
        public void PublicForeground(SkinElementInfo ee) {
            base.DrawSkinForeground(ee);
        }
    }
}
